omega/omega-connector/omega-connector-grpc/src/main/java/org/apache/servicecomb/saga/omega/connector/grpc/LoadBalancedClusterMessageSender.java

```diff
@@ -0,0 +1,140 @@
+
+import static java.util.Collections.emptyList;
+
+import java.lang.invoke.MethodHandles;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.function.Consumer;
+
+import org.apache.servicecomb.saga.omega.context.ServiceConfig;
+import org.apache.servicecomb.saga.omega.transaction.MessageDeserializer;
+import org.apache.servicecomb.saga.omega.transaction.MessageHandler;
+import org.apache.servicecomb.saga.omega.transaction.MessageSender;
+import org.apache.servicecomb.saga.omega.transaction.MessageSerializer;
+import org.apache.servicecomb.saga.omega.transaction.TxEvent;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.grpc.ManagedChannel;
+import io.grpc.ManagedChannelBuilder;
+
+public class LoadBalancedClusterMessageSender implements MessageSender {
+  private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+  private final Map<MessageSender, Long> senders = new HashMap<>();
+  private final Collection<ManagedChannel> channels;
+
+  public LoadBalancedClusterMessageSender(String[] addresses,
+      MessageSerializer serializer,
+      MessageDeserializer deserializer,
+      ServiceConfig serviceConfig,
+      MessageHandler handler) {
+
+    if (addresses.length == 0) {
+      throw new IllegalArgumentException("No reachable cluster address provided");
+    }
+
+    channels = new ArrayList<>(addresses.length);
+    for (String address : addresses) {
+      ManagedChannel channel = ManagedChannelBuilder.forTarget(address)
+          .usePlaintext(true)
+          .build();
+
+      channels.add(channel);
+      senders.put(
+          new GrpcClientMessageSender(
+              address,
+              channel,
+              serializer,
+              deserializer,
+              serviceConfig,
+              handler),
+          0L);
+    }
+  }
+
+  LoadBalancedClusterMessageSender(MessageSender... messageSenders) {
+    for (MessageSender sender : messageSenders) {
+      senders.put(sender, 0L);
+    }
+    channels = emptyList();
+  }
+
+  @Override
+  public void onConnected() {
+    senders.keySet().forEach(sender -> {
+      try {
+        sender.onConnected();
+      } catch (Exception e) {
+        log.error("Failed connecting to alpha at {}", sender.target(), e);
+      }
+    });
+  }
+
+  @Override
+  public void onDisconnected() {
+    senders.keySet().forEach(sender -> {
+      try {
+        sender.onDisconnected();
+      } catch (Exception e) {
+        log.error("Failed disconnecting from alpha at {}", sender.target(), e);
+      }
+    });
+
+  }
+
+  @Override
+  public void close() {
+    channels.forEach(ManagedChannel::shutdownNow);
+  }
+
+  @Override
+  public void send(TxEvent event) {
+    boolean success = false;
+    do {
+      try {
+        withFastestSender(messageSender -> {
```

Comment:
I think we should add sleep for it if the message is not sent out rightly.

Reply:
doing so will slow down the entire transaction...

Reply:
If the sender always failed to send  the message,  the CPU usage could be very high.

Reply:
that's what https://issues.apache.org/jira/browse/SCB-211 is for due to entire cluster unreachable